### DISCUSSION TITLE
Internal mechanics

### DISCUSSION THREAD

**Discussion by mathieu-pousse** [keycloakify/keycloakify#432] [General]
Hi there, 

I'm trying to debug / understand how does this project work.
I've read with attention https://github.com/keycloakify/keycloakify/discussions/346#discussioncomment-5889791 but one question remained unanswered.

I cannot figure out where does the data_model comes from. 
It is a built-in freemarker variable but I'm wondering what it does contain and where it is set in Keycloak.

I have an issue with the account theme and the  /applications page that takes ages to generate (keycloakify 8.1.2 / KC v21.1.1) 
It seems to introspect way too much information: 

Things are going bad under `applications.applications.realmRolesAvailable[].container.masterAdminClient...` because it reaches the limit:

![image](https://github.com/keycloakify/keycloakify/assets/2642023/d0348236-f2ec-43c0-b0b8-09fcc4e23dad)

If I can understand the object used for the serialization, I might get more in details with this issue :pray: 

Thanks in advance
M.

**Comment by garronej** [comment]
Hey @mathieu-pousse thanks for reaching out,  

So what's the problem exactly? There are some values that you need and that are not available in the kcContext or is it just that you are worried by the `/* too many recusive calls */`?

**Reply by mathieu-pousse** [reply]
The problem is that  it seems the introspection and serialization of the data_model is too long.

I suspect that somehow, the ftl is trying to serialize a too big object that causes a lot of java introscpection and that eventually takes ages to generate the javascript object.

It ends up with a page of 12Mb :open_mouth: 

That's why I'd like to know where the source object `data_model` comes from to get more in details and figure out if the template should exclude some properties or if this is due to a misconfiguration in my KC

**Reply by garronej** [reply]
@mathieu-pousse,  
12Mb!! Yeah, that's obviously a problem.  
As you might have understand, Keycloakify blindely pass over the ftl context to the frontend.  
When there are known values that cant easily be serialized we create a exception rule to excludes thoses.  
We have [quite a few of theses rules in place](https://github.com/keycloakify/keycloakify/blob/eb64fe60d06238a79e0e0969f062fe48ccfde3bb/src/bin/keycloakify/generateFtl/ftl_object_to_js_code_declaring_an_object.ftl#L446-L485), if you can provide me with a reproduction path I can include one for you issue.  

There is another way however. From Keycloakify v9 we're going to include the java code related to the account within the jar that keycloakify generates. This means that we'll be able to fix potential issue on the java side without having to go though the process of getting it merged on the Keycloak side.  
You can look at the java code here: https://github.com/keycloakify/keycloakify/tree/v9/src/bin/keycloakify/generateJavaStackFiles/account-v1-java  
It's an exact copy of the account code that comes with Keycloak 21.  

Let me know how I can help to sort out this issue.

**Reply by mathieu-pousse** [reply]
OK, I see but I'm wondering if we can simply drop the "applications" object from the payload of the application page (https://xxx/auth/realms/master/account/applications)  :sweat_smile: 

It will save a lot of effort for the page generation but I have no clue of the impact on existing themes :shrug: 

This page returns the list of client and in include all the realm description.
![image](https://github.com/keycloakify/keycloakify/assets/2642023/9042b123-c720-4736-8964-24c0c829fccf)

Do you have an idea of the usage that is made of this data on the page (in general I mean) ?

**Reply by garronej** [reply]
Seems like the applications object is used in the default template, I'm sure there's some nested properties that can be removed but we cannot remove `kcContext.applications` alltogether.  

https://github.com/keycloak/keycloak/blob/21.1.2/themes/src/main/resources/theme/base/account/applications.ftl

Can you sugest a pattern of kcContext path that you think could be removed?

**Reply by mathieu-pousse** [reply]
I would recommend removing 

- `applications.applications[].client.realm.*` 
- `applications.applications[].**.masterAdminClient`

This does not seem to be used on the templates and should make a good progress :ok_hand: 

Thanks for that :pray:

**Reply by garronej** [reply]
Ok! I put this in my backlog for this week end.  
I'll spare time to work on Keycloakify.

**Reply by garronej** [reply]
@mathieu-pousse I've released a candidate: 8.2.2-rc.0  
Can you test it and tell me if it fixes things?  
> applications.applications[].**.masterAdminClient  

About this path, I don't have the tooling to match this pattern. Doing things in FreeMarker is much harder to write and to test.  
As an alternative I have excluded everything on the page `application.ftl` that match the path `**.masterAdminClient`.  
The string `masterAdminClient` seems generic enough that it should not drop important field... I guess?  

This are the relevent changes:  

https://github.com/keycloakify/keycloakify/pull/440/files

**Reply by mathieu-pousse** [reply]
looks like a big improvement :ok_hand:

**Reply by garronej** [reply]
Pk, great! So you think I can release this version?

**Reply by mathieu-pousse** [reply]
lgtm

**Reply by garronej** [reply]
It's beeing released as 8.2.2 thanks a lot for your help on this @mathieu-pousse
